### ISSUE TITLE
Strange ProgID ?

### ISSUE THREAD

**Issue by Carlos-Rocha** [wqweto/UMMM#2]
Hello wqweto,

Using a manifest generated by ummm I successfully run my application (not a VB6 exe) in my Win8.1, but if I run it within Win7 (in a Virtual Machine) it doesn't run.

After inspecting the manifest I found that every ProgID for Codejock controls are not what I was expecting, eg instead of "Codejock.CommandBarsGlobalSettings.15.3.1" the manifest contains "XtremeCommandBars.CommandBarsGlobalSettings". So far this happens only with Codejock controls.

As I use ProgIDs to create controls I thought my application shouldn't run even in my Win8.1 machine, but it does. I'm sure the manifest is used in both cases.
The respective entry in .ini file for ummm is "File bin\cj\Codejock.CommandBars.Unicode.v15.3.1.ocx".

Is there anything I can do to make ummm generate the right ProgID?

Thanks,
Carlos

**Comment by wqweto** [comment]
I have no idea where VB6 is acquiring this different library name, but you should be ok creating controls with the "real" version independent ProgIds. I've seen this ProgiId disparity with other controls too -- MS Common Control 5.0 for instance. 
```
    Dim b As ComctlLib.ListView    
    Set b = Controls.Add("ComctlLib.ListView", "list")
```
`Controls.Add` generates this error:

![error](https://dl.unicontsoft.com/upload/pix/ss_comctl_error.png)

UMMM generates exactly `Comctl.ListViewCtl` from typelib info.

Can you check if there is `Codejock.CommandBarsGlobalSettings` ProgId in your registry under HKCR (doubt it). Last thing, is there a reason to use version-dependent ProgIDs?

**Comment by Carlos-Rocha** [comment]
In HKCR there is "Codejock.CommandBarsGlobalSettings.15.3.1".

I'm using "CA-Visual Objects", not VB, and to be able to dynamically create early-bounded controls it generates the wrapper classes. This is an automatic process and uses version-dependent ProgIDs. I don't know from where they get the ProgIDs, but I used to use an old tool  - Regsvr42 - that could generate the correct ProgIDs.

Anyway, I just checked in the manifest that there are controls with two ProgID, the second one inside <progid></progid>  tags, just before </comClass>. I'll test if this works for me.

Thanks

**Comment by Carlos-Rocha** [comment]
After looking around I think the declarations needed in your code are 
```
Private Type GUID
    Data1            As Long
    Data2            As Integer
    Data3            As Integer
    Data4(0 To 7)    As Byte
End Type

Private Declare Function ProgIDFromCLSID Lib "ole32.dll" (pCLSID As GUID, lpszProgID As Long) As Long
```



I tried to include it in your code but my knowledge of VB6 is very limited. Do you think it worth the trouble?

**Comment by wqweto** [comment]
You can try latest 1.0.7 build. I've included additional ProgIDs for controls based on typelib information. Not sure if this is the real fix for your issue.

**Comment by Carlos-Rocha** [comment]
Hi Vladimir,

I see no differences. The fact is that, as you already said, the result in the manifest is retrieved from the TypeLib, but it seems it's not enough to find the other ProgID (I'm not an expert, tho).
The ProgID needed by (My) VisualObjects and, I believe, by VB CreateObject() is the one retrieved from the registry - HKCR\Wow6432Node\CLSID. On 32bit OSs the path should be different, but I think both may be retrieved by ole32 function ProgIDFromCLSID.

I tried to understand your code inside pvDumpClasses, but I don't see how the .Guid is changed for each class within the For loop. It should be somewhere in that loop that a call to ProIDFromCLSID would provide the "other" ProgID...I think :).

I will keep trying, thanks

**Comment by wqweto** [comment]
The ProgID you are missing is completely bonkers. UMMM is already consulting registry under CLSID node for coclasses ProgID + VersionIndependentProgID which is what the `ProgIDFromCLSID` API does behind the scenes.

VB6 does not use this "alternative" ProgID you are missing. If it's used in `Controls.Add` you get the error I linked beforehand, including the *correct* ProgID in the error message, the one UMMM puts in the manifest.

I suspect that these "alternative" ProgIDs might be connected with licensing. Will have to research typelib interfaces in MSDN.

**Comment by wqweto** [comment]
I just implemented obtaining ProgID from API in commit 94be8622024d32193a1e3b0969c0470f55765a41

**Comment by dexwerx** [comment]
@Carlos-Rocha does this fix the original issue?

**Comment by Carlos-Rocha** [comment]
No, still generates "XtremeCommandBars.CommandBars".

But, despite working, UMMM writes down a critical error to the console (cmd.exe)

**critical: Method '~' of object '~' failed (pvDumpIdentity) (I see this sometimes with my own programs so it's probably something in my side).**

and a lot of warnings, like

**warning: coclass  GUID is duplicate of  ({E245C475-92F4-4AF7-9231-F158D1A035C3}) in bin\cj\Codejock.CommandBars.Unicode.v15.3.1.ocx"**

**Comment by wqweto** [comment]
@Carlos-Rocha It must be `m_oFSO.GetFileVersion` failing with a DLL/OCX. Will have to silence it as it's non fatal to default to 1.0.0.0.

I'm planning on installing a trial of Codejock's controls and figure out what's the trouble with this OCX but had no time for this recently.

**Comment by Carlos-Rocha** [comment]
You don't have to install the full trial, I'll send you CommandBars.ocx
Take your time :)

**Comment by dexwerx** [comment]
Interesting. I prefer the Version independant ProgIDs. Is there a reason to use the versioned ones?

**Comment by wqweto** [comment]
@dexwerx UMMM tries to include both version independent and versioned ProgIDs if present.

**Comment by wqweto** [comment]
@Carlos-Rocha It turns out the problem with Codejock controls was an obscure "feature" of typelib registration in registry, namely the version key is written *in hex* so "15.3" becomes "e.3" (fix in 12e74e9f20db54b03131e871212d9631144caec9) -- same familiar non-sense as in VB6 project files :-))

I tried to scour the net for info on `version` attribute on `<typelib>` nodes in manifest files but couldne't find anything conclusive. I'm writing version in decimal as of now but cannot confirm this is the correct behavior.

Let me know if this fixes the issue with you.

**Comment by Carlos-Rocha** [comment]
It does fix, thank you.
Codejock provided a tool to create SxS manifests for their controls, and the ProgIDs generated are the same.
